pkg/kubectl/apply/visitor.go

```diff
@@ -66,3 +66,8 @@ const (
 	// ReplaceStrategy is the strategy to replace the remote value with the local value
 	ReplaceStrategy = "replace"
 )
+
+// DoConflictDetector on strategy for element calling conflict method
+type DoConflictDetector interface {
```

Comment:
wondering why do we have this interface ?

Reply:
This interface is used for abstraction of defining a generic behavior on strategy level. It invokes type specific conflict method of element. Whatever merge strategy, replace strategy or delegate strategy has this kind of similar behavior. I want to keep the code style following the "visitor pattern" in merge and replace code base without breaking out. If conflict detection logic needs to be iterated in the future, the DoConflictDetector implement is the only place needs to be changed.

Reply:
It is used for an abstraction of a generic behavior on strategy level. This behavior invokes element's type specific conflict detection method for each coming element. All strategies, like merge and replace, will do the behavior in the similar way, so calling the interface to encapsulate the underlying implementation for each. Another benefit is that it retains the merge and replace strategy code base following "visitor" pattern. If there are iterations in conflict detection behaviors, the DoConflictDetector implementation is the only place needs to be changed.

Reply:
I guess the question is why you have this interface, while it's not used anywhere.

Reply:
Yes, interface abstraction is not necessary. Just define and implement method on strategy is sufficient